test/test_source.cpp

```diff
@@ -0,0 +1,187 @@
+
+        vector<KernelArg> args;
+        args.emplace_back(KernelArg{hX, bufferSize});
+        args.emplace_back(KernelArg{hY, bufferSize});
+        args.emplace_back(KernelArg{hOut, bufferSize, true});
+        args.emplace_back(KernelArg(&datasize));
+    
+        Headers headers_default;
+        headers_default.insert(Header{"cuda_runtime.h"});
+
+        //A2. Preparing the output for kernel-compilation
+        int *hostCompareOutput_Controlled = new int[5]{7,9,11,13,15};
+        int *hostCompareOutput_1 = new int[5]{12,14,16,18,20};
+        int *hostCompareOutput_2 = new int[5]{6,14,24,36,50};
+
+        Source source{
+            "#include \"cuda_runtime.h\"\n"
+            "extern \"C\"\n"
+            "__global__ void cuda_add(int *x, int *y, int *out, int "
+            "datasize) {\n"
+            " int i = threadIdx.x;\n"
+            " out[i] = x[i] + y[i];\n"
+            "}", headers_default};
+            
+        dim3 grid(1);
+        dim3 block(5);
+        
+        /*A3. Compilation of Kernels through the creation of a 
+          program from kernel - source*/
+        source.program("cuda_add").compile().configure(grid, block).launch(args);
+
+        //A4. Comparing the results
+        for (int i=0;i<5;i++){
+            REQUIRE(hOut[i] == hostCompareOutput_Controlled[i]);
+        }
+
+        //B1. Use of more than one header, in which checking-mechanisms are simulated.
+        WHEN("More than one headers are used."){
+            Headers headersNew1;
+            headersNew1.insert(Header{"cuda_runtime.h"});
+            headersNew1.insert(Header{"test_compare.hpp"});
+           
+            /*B2. Compilation of Kernels through the creation of a 
+            program from kernel - source*/
+            THEN("The results are consistent with the given controlled output."){
+                Source sourceNew1{
+                    "#include \"cuda_runtime.h\"\n"
+                    "#include \"test_compare.hpp\"\n"
+                    "extern \"C\"\n"
+                    "__global__ void cuda_add_with_header(int *x, int *y, int *out, int"
+                    " datasize) {\n"
+                    " compare check = OUT_COMPARE_WRONG; \n"
+                    " int i = threadIdx.x;\n"
+                    " out[i] = x[i] + y[i] + 5;\n"
+                    " if(out[i]%5==0) check = CORRECT;\n"
+                    "}", headersNew1};
+
+                dim3 gridNew1(1);
+                dim3 blockNew1(5);
+                
+                sourceNew1.program("cuda_add_with_header").compile().configure(gridNew1, blockNew1).launch(args);
+                
+                //B3. Comparing the results
+                for (int i=0;i<5;i++) REQUIRE(hOut[i] == hostCompareOutput_1[i]);
+            }    
+        }
+
+        //C1. Checking for consistencies for running various functions in a kernel-code
+        WHEN("More one functions in a kernel-code are used."){
+            args.clear();
```

Comment:
do you need to reset the KernelArgs here?

Reply:
Denn ich würde gerne aus praktischem Grund das gleiche KernelArgs-Array für weitere Testfälle einsetzen. Dazu muss ich es aufräumen, sonst kommt es zu einem Fehler.